zeekpkg/_util.py

```diff
@@ -69,37 +70,80 @@ def make_symlink(target_path, link_path, force=True):
             raise error
 
 
-def safe_tarfile_extractall(tfile, destdir):
-    """Wrapper to tarfile.extractall(), checking for path traversal.
+def zkg_tarfile_create(basedir):
+    compression = "gz"
+    tar_name = "".join((basedir, ".tar.", compression))
+
+    with tarfile.open(tar_name, "w:" + compression) as tar:
+        tar.add(basedir, arcname=".", filter=zkg_tarfile_create_filter)
+
+    return tar_name
+
+
+def zkg_tarfile_extractall(tfile, destdir):
+    """Wrapper to tarfile.extractall() using our filter that calls data_filter.
 
-    This adds the safeguards the Python docs for tarfile.extractall warn about:
+    This adds a lot of sanity checking for the tar file.
 
-    Never extract archives from untrusted sources without prior inspection. It
-    is possible that files are created outside of path, e.g. members that have
-    absolute filenames starting with "/" or filenames with two dots "..".
+    A zeek package contains arbitrary code that will be executed with the same
+    permissions as the user running zeek. Never extract archives from untrusted sources.
 
     Args:
         tfile (str): the tar file to extract
 
         destdir (str): the destination directory into which to place contents
 
     Raises:
-        Exception: if the tarfile would extract outside destdir
+        Exception: see tarfile.data_filter, tarfile.TarFile.extractall
     """
 
-    def is_within_directory(directory, target):
-        abs_directory = os.path.abspath(directory)
-        abs_target = os.path.abspath(target)
-        prefix = os.path.commonprefix([abs_directory, abs_target])
-        return prefix == abs_directory
-
     with tarfile.open(tfile) as tar:
-        for member in tar.getmembers():
-            member_path = os.path.join(destdir, member.name)
-            if not is_within_directory(destdir, member_path):
-                raise Exception("attempted path traversal in tarfile")
+        tar.extractall(destdir, filter=zkg_tarfile_extract_filter)
+
+
+def zkg_update_perms(new_attrs, member, extract):
+    # we are doing our own thing with `mode` here
```

Comment:
This is just the inverse of what was discussed in the ticket:

If a user creates a tar by hand or with custom tooling that contains a 0600 file and extracts it as `zeek:zeek` user on the target system, today it'll be a somewhat expected `0600`. With this change the file will be world-readable on the target system.

Having a `0600` file with zkg/git may be practically difficult to achieve and I'm not sure there's an actual use case, but not sure putting a world-readable label on everything is great.

If we'd warn users about strict umask settings and the consequences and a possible opt-in to set laxer permissions within the bundle, maybe that would suffice, too?

Reply:
I think unless there is a really strong use-case against it, we should favor the solution that will work for everyone. I feel like the package manager should be in charge of the permissions rather than carrying over (often accidental) permissions from the system on which the bundle is created (more so if they are creating them by hand) -- especially since there are also unexpected interactions depending on the umask zkg runs with (the permissions of the unpacked files don't have any effect on compiled code, even with this PR). A package should be portable, and if the end-user needs to worry about permissions on both the bundling system and the unbundling system, I don't think the package is very portable.

What is the best practice for running zkg? It seems like it would be best for zkg to run as a user that is different from the user running zeek, as well as a non-root user. It would be great if we could mask permissions with say `0o027`, but that would require zkg to know a group that is shared between the user unpacking the bundle and the user running zeek (or at least for end users to know to set the zkg users primary group to something the zeek user is a member of). Unless zkg is going to be opinionated on users and groups (which I don't think it should), I think it would be better to default to world-readable permissions.

If we want to provide a way for users to control permissions of files written by zkg, maybe it would be better to provide a cli/config option for `umask`, and default it to 0o022? Then, we could set that as the umask for the zkg process early on, and also mask the permissions we are calculating.

The only use-case I have come up with for wanting some files to have more restrictive permissions than others is if a user is including sensitive files in a bundle, which I hope they are not -- however even this use-case could be handled by using a more restrictive umask for all files.

Reply:
> The only use-case I have come up with for wanting some files to have more restrictive permissions than others is if a user is including sensitive files in a bundle, which I hope they are not 

Yeah, that was the only thought, and can also see that it's a "don't do that".

I think I'm still leaning on the side of putting the portable/appropriate permission bits into the bundle/tar, but during extraction at least not make them more permissible. Maybe I'm too timid and someone else should chime in.

Reply:
Would it be better to separate that piece into another ticket/PR?

Also, since the process umask is honored while building (for files created during the build step), should we also honor the process umask while unbundling to be more consistent? I'm not really sure what the best approach is there, but I really don't like the inconsistency.

Reply:
> Would it be better to separate that piece into another ticket/PR?

I'd wait and see if @ckreibich or others opine. As said, maybe I'm just too timid and making everything 0644/0755 is okay.

> should we also honor the process umask while unbundling to be more consistent?

Hmm, with the current state of the PR, I'd even say you could open-up/force the umask to `022` before building so that compiled files are world readable. In your test, the `.so` file produced has `0750`, which wouldn't be usable for non-root or non-zkg-user.
```
-rwxr-x--- root/root     31448 2024-07-25 16:44 Demo_Rot13/lib/Demo-Rot13.linux-x86_64.so
```

And even with the stricter permission handling, maybe a warning is appropriate if umask is found to be too strict before building.

Reply:
The latest commit (1ff7c78) should set the umask on build based on a new `zkg_umask` attribute on the Manager class, as well as applying that umask to files as they are extracted from a bundle. If we want to keep it, I think we could make `zkg_umask` user-configurable. It is currently hardcoded to `0o022`, but it seems to behave as I expected when changing that mask.